apm-agent-plugins/apm-apache-httpclient/apm-apache-httpclient4-plugin/src/main/java/co/elastic/apm/agent/httpclient/v4/ApacheHttpAsyncClientInstrumentation.java

```diff
@@ -89,42 +89,37 @@ public static class ApacheHttpAsyncClientAdvice {
+
+            ElasticContext<?> parentContext = tracer.currentContext();
+            if (parentContext.isEmpty()) {
+                // performance optimization, no need to wrap if we have nothing to propagate
+                // empty context means also we will not create an exit span
                 return null;
             }
-            Span<?> span = parent.createExitSpan();
-            HttpAsyncRequestProducer wrappedProducer = requestProducer;
             FutureCallback<?> wrappedFutureCallback = futureCallback;
-            boolean responseFutureWrapped = false;
-            if (span != null) {
-                span.withType(HttpClientHelper.EXTERNAL_TYPE)
-                    .withSubtype(HttpClientHelper.HTTP_SUBTYPE)
-                    .withSync(false)
-                    .activate();
-
-                wrappedProducer = asyncHelper.wrapRequestProducer(requestProducer, span, null);
-                wrappedFutureCallback = asyncHelper.wrapFutureCallback(futureCallback, context, span);
-                responseFutureWrapped = true;
-            } else {
-                wrappedProducer = asyncHelper.wrapRequestProducer(requestProducer, null, parent);
+            Span<?> span = null;
+            ElasticContext<?> parent = tracer.currentContext();
+            if (parent.getSpan() != null) {
+                span = parent.createExitSpan();
+                if (span != null) {
+                    span.withType(HttpClientHelper.EXTERNAL_TYPE)
+                        .withSubtype(HttpClientHelper.HTTP_SUBTYPE)
+                        .withSync(false)
+                        .activate();
+                    wrappedFutureCallback = asyncHelper.wrapFutureCallback(futureCallback, context, span);
+                }
             }
-            return new Object[]{wrappedProducer, wrappedFutureCallback, responseFutureWrapped, span};
+            HttpAsyncRequestProducer wrappedProducer = asyncHelper.wrapRequestProducer(requestProducer, span, tracer.currentContext());
+            return new Object[]{wrappedProducer, wrappedFutureCallback, span};
         }
 
         @Advice.OnMethodExit(suppress = Throwable.class, onThrowable = Throwable.class, inline = false)
         public static void onAfterExecute(@Advice.Enter @Nullable Object[] enter,
                                           @Advice.Thrown @Nullable Throwable t) {
-            Span<?> span = enter != null ? (Span<?>) enter[3] : null;
+            Span<?> span = enter != null ? (Span<?>) enter[2] : null;
             if (span != null) {
                 // Deactivate in this thread. Span will be ended and reported by the listener
                 span.deactivate();
-
-                if (!((Boolean) enter[2])) {
-                    // Listener is not wrapped- we need to end the span so to avoid leak and report error if occurred during method invocation
-                    span.captureException(t);
-                    span.end();
-                }
```

Comment:
instead of removing this we should probably test for `t != null` if an exception was thrown in the instrumented method.

Reply:
Good idea to add a test, checking this kind of corner cases with coverage is usually helpful !